### PR TITLE
Connection fixes

### PR DESCRIPTION
**Pull request by vspinu**
A range of connection related fixes. One commit per issue. 

I will fix the last issue regarding auto-association to projects and jars in another day or two.

### CODE REVIEW

File: cider.el [hunk 1]
```diff
@@ -1208,6 +1215,29 @@ non-nil, don't start if ClojureScript requirements are not met."
...
+                              (let ((ses-params (cider--gather-session-params ses)))
+                                (and (equal proj-dir (plist-get ses-params :project-dir))
+                                     (or (null port)
+                                         (equal port (plist-get ses-params :port)))
+                                     (or (null host)
+                                         (equal host (plist-get ses-params :host))))))
+                            sessions)))
+
+    (when session
+      (unless (y-or-n-p "A session with the same parameters exists.  You can connect a sibling instead.  Proceed? ")
```
Comment (bbatsov) [inline cider.el]: Maybe you can also print `session` as part of the message?

File: cider.el [hunk 2]
```diff
@@ -1208,6 +1215,29 @@ non-nil, don't start if ClojureScript requirements are not met."
                               cider-offer-to-open-cljs-app-in-browser)
                      (cider--offer-to-open-app-in-browser nrepl-server-buffer)))))))
 
+(defun cider--check-same-kind-session (params)
```
Comment (bbatsov) [inline cider.el]: I'm not sure if `kind` is the right word for this. Maybe `check-for-existing` or `check-for-similar`?

File: cider-connection.el [hunk 3]
```diff
@@ -639,6 +639,17 @@ function with the repl buffer set as current."
 
 ;;; Current/other REPLs
 
+(defun cider--no-repls-user-error (type)
+  "Throw \"No REPL\" user error customized for TYPE."
+  (let ((type (cond
```
Comment (bbatsov) [inline cider-connection.el]: Probably `pcase` would simplify this code. Or `cl-case`.

File: cider.el (hunk 1 above)
Comment (vspinu) [reply cider.el]: Could be, just a name though. Putting the full session is an overkill; too much stuff.

**Comment by bbatsov** [comment]
Overall the changes look good to me. Probably some simple test for the first and the last fix would be handy to make sure we don't accidentally break this down the road. 

I'm not sure I like the term `pending-cljs`, but I can't suggest something better of the top of my head. I'd like some term that suggests better that some REPL upgrade is in progress. I'm also wondering how are going to recover REPLs that for some reason don't managed to complete their "upgrade".

File: cider.el (hunk 2 above)
Comment (vspinu) [reply cider.el]: `check-for-existing` sound right to me.

File: cider.el (hunk 1 above)
Comment (bbatsov) [reply cider.el]: 👍

File: cider.el (hunk 2 above)
Comment (bbatsov) [reply cider.el]: 👍

**Comment by vspinu** [comment]
> I'm not sure I like the term pending-cljs,

Yeh. I actually did some digging on this one with synonyms et. al. But nothing better came up. I started with `uninitialized-cljs` but that's way too long for the modeline. 

> I'm also wondering how are going to recover REPLs that for some reason don't managed to complete their "upgrade".

We don't need to. If upgrade failed the repl returns to clj automatically on the first state message.

File: cider-connection.el (hunk 3 above)
Comment (vspinu) [reply cider-connection.el]: :) I don't see how it could be even simpler than that, it's a bare-bones conditional already.

**Comment by vspinu** [comment]
I have addressed the comments and added a test for the last commit. Writing tests for the first issue is quite pointless. One would need to overwrite every single function in `cider-connect-sibling-cljs`, thus tightly coupling the test with the implementation.

File: cider-connection.el [hunk 4]
```diff
@@ -92,6 +92,21 @@ PARAMS is a plist containing :host, :port, :server and other parameters for
...
+  "Return server buffer for SESSION or nil if there is no server."
+  (seq-some (lambda (r)
+              (buffer-local-value 'nrepl-server-buffer r))
+            (cdr session)))
+
+(defun cider--gather-session-params (session)
+  "Gather all params for a SESSION."
+  (let (params)
+    (dolist (repl (cdr session))
+      (setq params (cider--gather-connect-params params repl)))
```
Comment (dpsutton) [inline cider-connection.el]: am i misreading this? This is a dolist and `setq` without accumulating previous values? Should this be `add-to-list` or `(setq params (cons params ...`?

Comment (dpsutton) [reply cider-connection.el]: i missed `params` being passed into that function. ignore me :)

**Pull request merged**